Feature/auth

zachferland opened the pull request.

pelle commented on src/Storage.js:
```diff
@@ -0,0 +1,69 @@
+    return this.del(this.challengeKey(pairId))
+  }
+
+  writeResponse(pairId, response) {
```
@zachferland I'm a bit uncertain of the need for both storing challenges and responses. Shouldn't it only be necessary for challenges?

zachferland replied:
Storing responses is for the desktop browser flow, similar to the messaging server. Once we we make the messaging server available, developers could deploy there own messaging server to use along with this server (or maybe want it on a single server). Mainly we can't use the uPort messaging server to these types of responses since anyone controlling the messaging server could could use any/all of these response to login at a given service as a given user.

pelle commented on examples/AuthServer.js:
```diff
@@ -0,0 +1,51 @@
+  })
+})
+
+app.get('/authresponse/:pairID', (req, res) => {
```
@zachferland the browser should be responsible for interacting with it's backend server. There are so many different kinds of backends that people will use and we just need to solve the generic flow similar to what oauth2 does. Then they can handle the backend part.

zachferland replied:
@pelle yeah I am okay with sending it back the browser. Sending the response back to the server would be stronger model, since it minimizes (almost entirely) the chance that it is leaked. The token (and/or additional information) could be leaked. For example on mobile when passing from app to app (uport to browser, maybe another app captures it), or capturing in browser history since in the url. But yes this risk seems minimal right now and passing it to the browser would help keep the same flows were already have in mobile and in the libraries.

zachferland replied:
@pelle actually on second thought it may make more sense to send all responses back to the server which are responses for requests generated on the same server. There are a few benefits, 1) it would always allow the flow to be the same from the perspective of the server (instead of sometimes the server has to write the response to storage, other times it has to immediately return the response, its not clear how to differentiate these), 2) it would allow the flow to be more easily managed in the library and wrapped up for end users, it could always return the server response  as a result of valid/invalid CR (the alternative would be, when in mobile browser the lib would get the response from mobile of the challenge (then have to get the server response), while on desktop the lib would get the server response (not the mobile response, alternatively getting the mobile response would not make much sense here, as it would have been sent to the server then returned here by polling, then sent back to the server)) 3) may allow us to wrap up certain best practices in the libraries, like have a certain amount of control over how the response to the browser is paired to the challenge/response. Ultimately either way a developer can do as they please, but this would likely improve the experience in the libraries.

pelle commented on src/Storage.js:
```diff
@@ -0,0 +1,69 @@
+import redis from 'redis'
+
+class Storage {
```
We should indicate the functions that need to be implemented. I think maybe renaming it ChallengeStorage is the best way and having simple (`get`, `set`, `clear`) functions.

zachferland replied:
yeah could improve this, and some additional docs on implementing your own storage interface should definitely be provided

zachferland commented:
thanks for the review @pelle. Going to add additional details/simplify the storage layer, change functionality to assume the response is passed to the browser, change minor things to align with mobile (challenge -> chl, etc), and confirm the examples work with mobile

pelle commented on src/AuthCredentials.js:
```diff
@@ -0,0 +1,82 @@
+        response.credentials = credentials
+        return this.responseStorage.set(pairId, JSON.stringify(response))
+      }
+      this.responseStorage.set(pairId, 'Error')
```
How response is returned to client is out of scope for this library. This should just return the response.

Eliminate the responseStorage. An example can be implemented in the example code

zachferland replied:
yeah somewhat, responseStorage is planned to be optional, so should only set if there is responseStorage, going to fix that

zachferland commented on __tests__/Credentials-test.js:
```diff
@@ -1,182 +0,0 @@
-import Credentials from '../src/Credentials'
```
yeah appears that way

zachferland commented on src/Storage.js:
```diff
@@ -0,0 +1,42 @@
+
+  set(key, value) {
+    return new Promise((resolve, reject) => {
+      this.client.set(this.key(key), value, (err, res) => {
```
perfect, yeah wanted to set the expire time

zachferland commented on src/Storage.js:
```diff
@@ -0,0 +1,42 @@
+    })
+  }
+
+  del(key) {
```
yeah its possible, but I added del such that it the storage interface is more clear for anyone looking to implement their own. I also don't think the high level code would expect it to also delete when you call get

zachferland commented on src/AuthCredentials.js:
```diff
@@ -0,0 +1,82 @@
+    return Promise.all([this.random(), this.random()]).then(rand => {
+      pairId = rand[0]
+      challenge = rand[1]
+      payload.challenge = `${pairId}.${challenge}`
```
yeah I agree with moving challenge/auth code into a separate class, definitely possible and useful. I think I would still use a AuthCredentials class though, rather than adding more if lines based on configs, auth is a clear use case that is likely better defined by having its own class, it may also expand, and be usefully defined in some cases/context as providing auth alone.

was also looking to consider this another PR/story since this one has gotten so large

pelle commented on README.md:
```diff
@@ -58,6 +58,79 @@ The networks object includes a set of networks for which JWTs will be verified o
 
 Look in [uport-lite](https://github.com/uport-project/uport-lite) for the default networks and registries which will be queried for JWT verification.
 
+## Authentication
```
I think we should probably change the wording here to be Enhanced  or Strong Authentication as regular Uport Connect also is Authentication.

pelle commented on src/AuthCredentials.js:
```diff
@@ -0,0 +1,88 @@
+  constructor (settings = {}) {
+    super(settings)
+    this.challengeStorage = settings.challengeStorage || new Storage('challenge')
+    this.responseStorage = settings.responseStorage || new Storage('response')
```
AuthCredentials should not use responseStorage. This makes too many assumptions about the implementation and integration into the consuming applications architecture.

pelle commented on src/AuthCredentials.js:
```diff
@@ -0,0 +1,88 @@
+    }).then(res => credentials)
+  }
+
+  getAuthResponse(pairId) {
```
`getAuthResponse` and `setAuthResponse` should not be part of this class

zachferland commented:
@pelle this is not the same response storage as before, earlier I had wrapped up the developer's server responses into some of the other functionality, which now has been removed (this is definitely better now). The response storage is just for our response from mobile, they are now separated from other functionality and are in  there own functions which you can choose to use or not. They simply serve the same purpose as the messaging server and were added to simplify having to explain that, so they are just light, completely optional functions to wrap some of that up. We already added this storage layer for challenges, it would be nice to use that to also provide the simply wrappers to set of the same functionality as the messaging server. It also helps to align with the implementation in uport-connect. In both libraries they are just one way to handle this flow which we provide by default, but in both libs they are entirely optional and we added the base parts that allow developers to implement  it how ever they please instead of using this default functionality.
